pkg/controller/workspacerouting/solvers/openshift_oauth_solver.go

```diff
@@ -0,0 +1,189 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package solvers
+
+import (
+	"fmt"
+	"github.com/che-incubator/che-workspace-operator/pkg/apis/workspace/v1alpha1"
+	"github.com/che-incubator/che-workspace-operator/pkg/common"
+	"github.com/che-incubator/che-workspace-operator/pkg/config"
+	routeV1 "github.com/openshift/api/route/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+type OpenShiftOAuthSolver struct{}
+
+var _ RoutingSolver = (*OpenShiftOAuthSolver)(nil)
+
+type proxyEndpoint struct {
+	machineName            string
+	upstreamEndpoint       v1alpha1.Endpoint
+	publicEndpoint         v1alpha1.Endpoint
+	publicEndpointHttpPort int64
+}
+
+func (s *OpenShiftOAuthSolver) GetSpecObjects(spec v1alpha1.WorkspaceRoutingSpec, workspaceMeta WorkspaceMetadata) RoutingObjects {
+	var exposedEndpoints = map[string][]v1alpha1.ExposedEndpoint{}
+	proxy, noProxy := getProxiedEndpoints(spec)
+	defaultIngresses, defaultEndpoints := getIngressesForSpec(noProxy, workspaceMeta)
+	for machineName, machineEndpoints := range defaultEndpoints {
+		exposedEndpoints[machineName] = append(exposedEndpoints[machineName], machineEndpoints...)
+	}
+
+	portMappings := getProxyEndpointMappings(proxy, workspaceMeta)
+	var proxyPorts = map[string][]v1alpha1.Endpoint{}
+	for _, proxyEndpoint := range portMappings {
+		proxyPorts[proxyEndpoint.machineName] = append(proxyPorts[proxyEndpoint.machineName], proxyEndpoint.publicEndpoint)
+	}
+	for machineName, machineEndpoints := range noProxy {
+		proxyPorts[machineName] = append(proxyPorts[machineName], machineEndpoints...)
+	}
+	// Use common service for all unproxied endpoints
+	proxyServices := getServicesForEndpoints(proxyPorts, workspaceMeta)
+	for idx := range proxyServices {
+		proxyServices[idx].Annotations = map[string]string{
+			"service.alpha.openshift.io/serving-cert-secret-name": "proxy-tls",
```

Comment:
P.S. Controller does not seem to update this field after reconciling, not sure it's expected
![Screenshot_20200406_133853](https://user-images.githubusercontent.com/5887312/78550095-2aba1400-780c-11ea-8b9c-282c0bc2f7f1.png)

Reply:
Ah good catch, we currently ignore `ObjectMeta` when checking if a service needs to be updated. I'll fix this.

Reply:
Looking into it a bit, this is maybe an task for another issue -- we can't naively check that annotations match since some are added by the cluster, and I don't want to hard-code checking `service.alpha.openshift.io/serving-cert-secret-name` in the generic case without further discussion.

For now, this annotation should not change (unless you rollout a new operator that changes the semantics).